### PR TITLE
Unit tests refactor

### PR DESCRIPTION
**Pull request by aditighag**
Refactor interface implementation in fakes

Embed interface into a mock struct so that all the methods
are inherited, and only the methods that are relevant to
unit test cases can be implemented to override actual code behavior.

This gets rid of redundant/no-op method implementations in fakes/mocks like -

type (m mockFoo) testFoo {
panic("Implement me")
}

type (m mockBar) testBar {

}

There are other places in the code which can use a similar kind of refactor, and can be addressed in a follow-up PR. Some examples - 
https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon_test.go#L152
https://github.com/cilium/cilium/blob/master/pkg/node/manager/manager_test.go

Signed-off-by: Aditi Ghag <aditi@cilium.io>

### CODE REVIEW

File: pkg/endpoint/redirect_test.go [hunk 1]
```diff
@@ -19,23 +19,21 @@ package endpoint
 import (
 	"context"
 
+	"gopkg.in/check.v1"
```
Comment (tklauser) [inline pkg/endpoint/redirect_test.go:22]: Minor nit: this import should go after the `github.com/cilium/cilium` imports into a separate group. We usually use the order stdlib, cilium, 3rd party for imports.

**Review by tklauser — APPROVED** [review]
Nice improvment! I didn't know this was possible to embed interface types to inherit methods. TIL.

One minor nit re. import ordering in case you're doing another revision.

**Comment by tklauser** [comment]
test-runtime

File: pkg/endpoint/redirect_test.go (hunk 1 above)
Comment (aditighag) [reply pkg/endpoint/redirect_test.go:22]: @tklauser Once the PR is ready to merge, I can fix the imports. Else, I'm anyway planning to do a follow-up so I fix it then.

**Review by jrajahalme — APPROVED** [review]
I wonder how informative the runtime errors will be if a code change leads to an "unimplemented" function the be called, as they would be called through the uninitialized (`nil`-valued) embedded interface?

**Comment by aditighag** [comment]
> I wonder how informative the runtime errors will be if a code change leads to an "unimplemented" function the be called, as they would be called through the uninitialized (`nil`-valued) embedded interface?

That's simple. The unit tests will fail immediately with "invalid memory..." trace.

**Comment by rolinh** [comment]
> > I wonder how informative the runtime errors will be if a code change leads to an "unimplemented" function the be called, as they would be called through the uninitialized (`nil`-valued) embedded interface?
> 
> That's simple. The unit tests will fail immediately with "invalid memory..." trace.

It will indeed fail with a `panic: runtime error: invalid memory address or nil pointer dereference` and the trace. It's not the most informative kind of error though. I think the approach we have for Hubble mocks is a little more friendly when it fails because a method is unimplemented:
```go
// FakeServiceGetter is used for unit tests that need ServiceGetter.
type FakeServiceGetter struct {
	OnGetServiceByAddr func(ip net.IP, port uint16) (service flowpb.Service, ok bool)
}

// GetServiceByAddr implements FakeServiceGetter.GetServiceByAddr.
func (f *FakeServiceGetter) GetServiceByAddr(ip net.IP, port uint16) (service flowpb.Service, ok bool) {
	if f.OnGetServiceByAddr != nil {
		return f.OnGetServiceByAddr(ip, port)
	}
	panic("OnGetServiceByAddr not set")
}
```

**Comment by kaworu** [comment]
> That's simple. The unit tests will fail immediately with "invalid memory..." trace.

IMHO given the trace it is far from obvious that the underlying issue is a mock misuse / missing implementation:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x499232]

goroutine 1 [running]:
main.(*SomeStruct).doSomething(0xc000010200, 0xc000068f58, 0x4dc020)
	<autogenerated>:1 +0x32
main.f(...)
	/tmp/sandbox588939945/prog.go:14
main.main()
	/tmp/sandbox588939945/prog.go:18 +0x59
```
[Even when looking at the source](https://play.golang.org/p/F-mT2B_rIEk) I find it difficult to debug compared to a `panic("SomeStruct.doSomething unimplemented")`.

**Review by tklauser — CHANGES_REQUESTED** [review]
> https://github.com/cilium/cilium/pull/14744#issuecomment-769676544
> https://github.com/cilium/cilium/pull/14744#issuecomment-769678332

These are fair points and I agree it might not be obvious from the trace that the issue is a missing mock/fake implementation. In my initial review I was naïvely assuming that we'd get compile time errors in these cases.

I think in this case the value of removing some empty implementations isn't worth the added complication for developers when adding/extending unit tests. I agree with @rolinh and @kaworu that the way Hubble implements these would be more developer-friendly.

**Review by aanm — APPROVED** [review]
nothing else to point out besides what Tobias said.

**Comment by aditighag** [comment]
@rolinh @kaworu Sure, that's a fair point. But at the same time it's not ideal to force developers to implement all the methods of an interface in unit tests. The hubble approach still doesn't solve this issue. We have quite a few broad interfaces in cilium agent (with ~5-10 methods), and writing no-op/"implement-me" forwarding methods for all of them becomes tedious, and leads to code bloat.

>In my initial review I was naïvely assuming that we'd get compile time errors in these cases.

@tklauser The Hubble approach doesn't fail at compile time either.

As a middle ground, how about we initialize the embedded interface with an actual struct that we are trying to mock out in unit test files? With this, developers can mock out test cases specific methods, other calls will go to the actual struct. 
Stack traces will clearly indicate if a call is going to the mock or actual struct.

**Comment by rolinh** [comment]
> at the same time it's not ideal to force developers to implement all the methods of an interface in unit tests. The hubble approach still doesn't solve this issue

I'm probably missing something but I think the hubble approach does solve the issue at hand, Take for instance this code snippet from `pkg/hubble/observer/local_observer_test.go`:

```go
func TestLocalObserverServer_OnGetFlows(t *testing.T) {
        // ... some code...
        fakeServer := &testutils.FakeGetFlowsServer{
                OnSend: func(response *observerpb.GetFlowsResponse) error {
                        assert.Equal(t, response.GetTime(), response.GetFlow().GetTime())
                        assert.Equal(t, response.GetNodeName(), response.GetFlow().GetNodeName())
                        flowsReceived++
                        return nil
                },
                FakeGRPCServerStream: &testutils.FakeGRPCServerStream{
                        OnContext: func() context.Context {
                                return context.Background()
                        },
                },
        }
        // ...some more code ...
        err = s.GetFlows(req, fakeServer)
        assert.NoError(t, err)
```

It makes use of this fake implementation:

```go
// FakeGetFlowsServer is used for unit tests and implements the
// observerpb.Observer_GetFlowsServer interface.
type FakeGetFlowsServer struct {
        OnSend func(response *observerpb.GetFlowsResponse) error
        *FakeGRPCServerStream
}

// Send implements observerpb.Observer_GetFlowsServer.Send.
func (s *FakeGetFlowsServer) Send(response *observerpb.GetFlowsResponse) error {
        if s.OnSend != nil {
                // TODO: completely convert this into using flowpb.Flow
                return s.OnSend(response)
        }
        panic("OnSend not set")
}
```

Only 2 methods of the 7 methods of the `observerpb.Observer_GetFlowsServer` interface had to be mocked for the test. If one of the other methods was required, say, `SendMsg(m interface{}) error`, then the test would panic with `"OnSendMsg not set"` as a message.

This approach means that the fakers need to be written once but then only what's required for the test need to be mocked.

> The Hubble approach doesn't fail at compile time either.

Yes but as pointed out above, it tells you which method of the "faker" need to be implemented for the test instead of a vague panic stack trace.

**Comment by rolinh** [comment]
> That's because the unit test struct embeds FakeGRPCServerStream, which in turn implements all the 7 methods. That's exactly the issue this PR is trying to avoid.

The methods are not really implemented as in they satisfy the interface but will panic if the correct func is not set for test. These "faker" are just helpers for testing. They are written once and then can be used in various tests. When writing tests, you start with the faker and then only implement the methods that are required for the test, again, as shown in [the example above](https://github.com/cilium/cilium/pull/14744#issuecomment-771914712).

> Can you explain why this approach isn't a good compromise? It solves the problem with code bloat/tedious method forwarding implementation, and at the same time gives you a user-friendly trace in case of failures.
> 
> > As a middle ground, how about we initialize the embedded interface with an actual struct that we are trying to mock out in unit test files? With this, developers can mock out test cases specific methods, other calls will go to the actual struct.
> > Stack traces will clearly indicate if a call is going to the mock or actual struct.

I'm not sure I understand the approach you're proposing. And I really don't see anything tedious nor bloat with the hubble approach so I probably don't understand what you are trying to solve.

**Comment by tklauser** [comment]
> @rolinh @kaworu Sure, that's a fair point. But at the same time it's not ideal to force developers to implement all the methods of an interface in unit tests. The hubble approach still doesn't solve this issue. We have quite a few broad interfaces in cilium agent (with ~5-10 methods), and writing no-op/"implement-me" forwarding methods for all of them becomes tedious, and leads to code bloat.
> 
> > In my initial review I was naïvely assuming that we'd get compile time errors in these cases.
> 
> @tklauser The Hubble approach doesn't fail at compile time either.

Yes, sorry. I wasn't trying to suggest it would fail at compile time with the Hubble approach. I just wanted to point out that if there is no solution with a compile time failure (as I initially thought the solution proposed in this PR would), the Hubble approach is more readable and easier to follow than embedding interfaces - even if it comes at the price of some added code which isn't used by tests. FWIW, there are even tools such as https://github.com/josharian/impl which will generate the stub methods for you.

> As a middle ground, how about we initialize the embedded interface with an actual struct that we are trying to mock out in unit test files? With this, developers can mock out test cases specific methods, other calls will go to the actual struct.
> Stack traces will clearly indicate if a call is going to the mock or actual struct.

I think this solution makes the tests much harder to follow as you need to jump to the concrete implementations embedded in the mocks/fakes in order to check which methods are implemented. And it mixes implementations into the tests which aren't necessarily part of the system under test.

The only benefit of this PR I can currently see is that it saves a few lines, by getting rid of a few harmless dummy methods. Is there any other advantage that I didn't see?

**Comment by tklauser** [comment]
FWIW: now that #14745 is merged, the `pkg/endpoint` fakes/mocks touched in this PR are gone altogether 🎉

**Comment by kaworu** [comment]
> @rolinh @kaworu Sure, that's a fair point. But at the same time it's not ideal to force developers to implement all the methods of an interface in unit tests. The hubble approach still doesn't solve this issue.

I agree.

> We have quite a few broad interfaces in cilium agent (with ~5-10 methods), and writing no-op/"implement-me" forwarding methods for all of them becomes tedious, and leads to code bloat.

I'm more scared of a few complex lines of code than a thousand obviously correct ones, but that's just my two cents.

> > In my initial review I was naïvely assuming that we'd get compile time errors in these cases.
> 
> @tklauser The Hubble approach doesn't fail at compile time either.

That is correct. The Hubble approach is very flexible with respect to mocking different behaviour, but that seems out of the scope of this PR to me.

> As a middle ground, how about we initialize the embedded interface with an actual struct that we are trying to mock out in unit test files? With this, developers can mock out test cases specific methods, other calls will go to the actual struct.
> Stack traces will clearly indicate if a call is going to the mock or actual struct.

That sounds promising but I'm unsure to fully understand the proposition, could you give an short code example?

To summarize, my understanding is the following: as-is this PR reduces code bloat using a nice trick (didn't know it was possible to embed interface, thanks @aditighag for teaching me this!), and it was pointed out that it was at the expense of the error message readability when an unimplemented method is called (versus a mock `panic` implementation). So it's a trade-off. Personally, I'm not feeling strongly about this.

**Comment by joestringer** [comment]
Marking as "draft" since it looks like this needs a rebase. When you would like additional review, scroll to the bottom of the page and click the "Ready for review" button.
